### Grid Search CV 
#### Defination :
1. GridSearchCV (Grid Search with Cross-Validation) is a method to find the best combination of hyperparameters for a machine learning model.

2. It systematically tries all possible combinations of parameters you give, and selects the one that gives the best performance, based on cross-validation.

#### Why use GridSearchCV? :
1. Manually testing hyperparameters is inefficient and error-prone.
2. Grid Search automates the process.
3. It uses cross-validation to reduce overfitting and ensure the model generalizes well.

#### Working :
1. we r using SVM model and want to find the best:
C: regularization parameter, kernel: 'linear' or 'rbf'

2. GridSearchCV will : 
- try all combinations of C and kernel
- train the model on trainig data using cross-validation
- score each combination
- return the best performing parameters

--- 

### KNN 
#### Defination : 
1. K-Nearest Neighbors (KNN) is a simple and effective machine learning algorithm used for classification and regression.
2. It works by finding the 'k' closest data points (neighbors) and using them to predict the output for a new data point.

#### Performance :
- Performance of KNN depends on hyperparameters like :
1. n_neighbors → Number of neighbors (k)
2. weights → Uniform or distance-based weighting
3. metric → Distance metric (e.g., 'euclidean', 'manhattan')

#### Why use GridSearchCV with KNN? :
- GridSearhCv helps in automatically finding the best values of these hyperparameters by :
1. Trying multiple combinations of k, weights, and metric
2. Evaluating each combination using cross-validation
3. Returning the combination that gives the best accuracy (or chosen score)



In [6]:
# Example : GridSearchCv with KNN 

from sklearn.datasets import load_iris
from sklearn.neighbors import KNeighborsClassifier
from sklearn.model_selection import GridSearchCV

#Load data
X, y = load_iris(return_X_y=True)

#define parameters grid
param_grid = {
    'n_neighbors': [3, 5, 7],
    'weights': ['uniform', 'distance'],
    'metric': ['euclidean', 'manhattan']  
}

# Create KNN model
knn = KNeighborsClassifier()

# Create GridSearchCV
grid_search = GridSearchCV(knn, param_grid, cv=5, scoring='accuracy')
grid_search.fit(X, y)

# Results
print("Best Parameters:", grid_search.best_params_)
print("Best Score:", grid_search.best_score_)

Best Parameters: {'metric': 'euclidean', 'n_neighbors': 7, 'weights': 'uniform'}
Best Score: 0.9800000000000001


This means that the optimal KNN model uses 7 neighbors, uniform weighting, and Euclidean distance, achieving 98% accuracy on the Iris dataset using 5-fold cross-validation.
